🌐 Add Ukrainian translation for `docs/uk/docs/tutorial/path-params.md` (#4570)
fastapi/fastapi: docs/uk/docs/tutorial/path-params.md

# Параметри шляху

> Ви можете оголосити "параметри" шляху або "змінні" з тим самим синтаксисом, що використовується строками Python:

Reviewer: ```suggestion
Ви можете оголосити "параметри" шляху або "змінні" з тим самим синтаксисом, що використовується стрічками Python:
```

> Ви можете оголосити "параметри" шляху або "змінні" з тим самим синтаксисом, що використовується строками Python:

Reviewer 2: ```suggestion
Ви можете оголосити "параметри" шляху або "змінні" з тим самим синтаксисом, що використовується рядками Python:
```

```Python hl_lines="6-7"
{!../../../docs_src/path_params/tutorial001.py!}
```

Значення параметра шляху `item_id` буде передано вашій функції як аргумент `item_id`.

Отже, якщо ви запустите цей приклад і перейдете до <a href="http://127.0.0.1:8000/items/foo" class="external-link" target="_blank">http://127.0.0.1:8000/items/foo</a>, ви побачите відповідь:

```JSON
{"item_id":"foo"}
```

## Параметри шляху з типами

Ви можете оголосити тип параметра шляху у функції, використовуючи стандартні анотації типу Python:

```Python hl_lines="7"
{!../../../docs_src/path_params/tutorial002.py!}
```

У цьому випадку `item_id` оголошується як `int`.

!!! check

>     Це надасть вам підтримку редактора всередині вашої функції, з перевіркою помилок, автозаповненням, тощо.

Reviewer: ```suggestion
    Це надасть вам підтримку редактора всередині вашої функції, з перевіркою помилок, автозаповненням тощо.
```

```JSON
{
    "detail": [
        {
            "loc": [
                "path",
                "item_id"
            ],
            "msg": "value is not a valid integer",
            "type": "type_error.integer"
        }
    ]
}
```

оскільки параметр шляху `item_id` мав значення `"foo"`, яке не є `int`.

Така ж помилка з’явилася б, якби ви вказали `float` замість `int`, як у: <a href="http://127.0.0.1:8000/items/4.2" class="external-link" target="_blank">http://127.0.0.1:8000/items/4.2</a>

!!! check
    Отже, з тим самим оголошенням типу Python, **FastAPI** надає вам перевірку даних.

    Зверніть увагу, що помилка також чітко вказує, де перевірка не пройшла.

    Це надзвичайно корисно під час розробки та налагодження коду, який взаємодіє з вашим API.

## Документація

А коли ви відкриєте браузер на <a href="http://127.0.0.1:8000/docs" class="external-link" target="_blank">http://127.0.0.1:8000/docs</a>, ви побачите автоматичну, інтерактивну документацію API, як-от:

<img src="/img/tutorial/path-params/image01.png">

!!! check

>     Знову ж таки, лише з тим самим оголошенням типу Python, **FastAPI** надає вам автоматичну інтерактивну документацію (інтегруючи інтерфейс користувача Swagger).

Reviewer: ```suggestion
    Знову ж таки, лише з тим самим оголошенням типу Python, **FastAPI** надає вам автоматичну інтерактивну документацію (інтегруючи Swagger UI).
```

Ви можете використовувати ті самі оголошення типу з `str`, `float`, `bool` та багатьма іншими складними типами даних.

Деякі з них розглядаються в наступних розділах туторіалу.

## Порядок має значення

Створюючи *операції шляху*, ви можете потрапити у ситуацію, коли у вас є фіксований шлях.

Як `/users/me`, скажімо, що це для отримання даних про поточного користувача.

І тоді ви також можете мати шлях `/users/{user_id}` для отримання даних про певного користувача за деяким ідентифікатором користувача.

Оскільки *операції шляху* визначаються в порядку, вам потрібно переконатися, що шлях для `/users/me` оголошено перед `/users/{user_id}`:

```Python hl_lines="6  11"
{!../../../docs_src/path_params/tutorial003.py!}
```

Інакше шлях для `/users/{user_id}` відповідатиме також для `/users/me`, "думаючи", що він отримує параметр `user_id` зі значенням `"me"`.

## Попередньо визначені значення

Якщо у вас є *операція шляху*, яка отримує *параметр шляху*, але ви хочете, щоб можливі дійсні значення *параметру шляху* були попередньо визначені, ви можете використовувати стандартний Python <abbr title="Enumeration (перерахування)">`Enum` (`перерахування`)</abbr>.

### Створіть клас `Enum`

> Імпортуйте `Enum` та створюйте підклас, який успадковується від `str` та від `Enum`.

Reviewer 3: ```suggestion
Імпортуйте `Enum` та створіть підклас, який успадковується від `str` та від `Enum`.
```

Успадкувавши від `str`, API документація знатиме, що значення мають бути типу `string` і вони зможуть правильно відображатися.

Потім створіть атрибути класу з фіксованими значеннями, які будуть доступними валідними значеннями:

```Python hl_lines="1  6-9"
{!../../../docs_src/path_params/tutorial005.py!}
```

!!! info
    <a href="https://docs.python.org/3/library/enum.html" class="external-link" target="_blank">Перерахування (`Enum`) доступні в Python</a> з версії 3.4.

!!! tip
    Якщо вам цікаво, «AlexNet», «ResNet» і «LeNet» — це лише назви <abbr title="Технічно, архітектури моделей глибокого навчання">моделей</abbr> машинного навчання.

### Оголосіть *параметр шляху*

Потім створіть *параметр шляху* з анотацією типу, використовуючи створений вами клас (`ModelName`):

```Python hl_lines="16"
{!../../../docs_src/path_params/tutorial005.py!}
```

### Перевірте документацію

> Оскільки доступні значення для *параметра шляху* попередньо визначені, інтерактивна документація може добре показати їх:

Reviewer: ```suggestion
Оскільки доступні значення для *параметра шляху* попередньо визначені, інтерактивна документація може показати їх:
```

<img src="/img/tutorial/path-params/image03.png">

### Робота з *перерахуванням* у Python

> Значенням *параметра шляху* буде *членом перерахування*.

Reviewer 3: ```suggestion
Значення *параметра шляху* буде *членом перерахування*.
```

Reviewer 2: @romashevchenko Тоді вже краще `членом переліку` або `членом перелічення`

```Python hl_lines="17"
{!../../../docs_src/path_params/tutorial005.py!}
```

#### Отримайте *значення перерахування*

Ви можете отримати фактичне значення (у цьому випадку `str`) за допомогою `model_name.value`, або загалом, `your_enum_member.value`:

```Python hl_lines="20"
{!../../../docs_src/path_params/tutorial005.py!}
```

!!! tip
    Ви також можете отримати доступ до значення `"lenet"` за допомогою `ModelName.lenet.value`.

#### Поверніть *члени перерахування*

Ви можете повернути *члени перерахування* з *операції шляху*, навіть вкладених у тіло JSON (наприклад, `dict`).

Вони будуть перетворені у відповідні значення (у цьому випадку рядки) перед поверненням їх клієнту:

```Python hl_lines="18  21  23"
{!../../../docs_src/path_params/tutorial005.py!}
```

У вашому клієнті ви отримаєте відповідь JSON:

```JSON
{
  "model_name": "alexnet",
  "message": "Deep Learning FTW!"
}
```

## Параметри шляху, що містять шляхи

Скажімо, у вас є *операція шляху* із шляхом `/files/{file_path}`.

Але вам потрібно, щоб сам `file_path` містив *шлях*, наприклад, `home/johndoe/myfile.txt`.

Отже, URL-адреса цього файлу буде схожна на: `/files/home/johndoe/myfile.txt`.

### Підтримка OpenAPI

> OpenAPI не підтримує спосіб оголосити *параметр шляху*, щоб він містив всередині *шлях*, оскільки це може призвести до сценаріїв, які важко перевірити та визначити.

Reviewer: ```suggestion
OpenAPI не підтримує спосіб оголошення *параметра шляху*, щоб він містив всередині *шлях*, оскільки це може призвести до сценаріїв, які важко перевірити та визначити.
```

Тим не менш, ви все ще можете зробити це в **FastAPI**, використовуючи один із внутрішніх інструментів від Starlette.

І документація все одно працюватиме, хоча не додасть жодної інформації, яка говорить про те, що параметр має містити шлях.

### Конвертор шляху

Використовуючи параметр безпосередньо зі Starlette, ви можете оголосити *параметр шляху*, що містить *шлях*, використовуючи URL-адресу, наприклад:

```
/files/{file_path:path}
```

У цьому випадку ім’я параметра – `file_path`, а остання частина, `:path`, повідомляє, що параметр має відповідати будь-якому *шляху*.

Отже, ви можете використовувати його з:

```Python hl_lines="6"
{!../../../docs_src/path_params/tutorial004.py!}
```

!!! tip
    Вам може знадобитися, щоб параметр містив `/home/johndoe/myfile.txt` з косою рискою (`/`).

    У цьому випадку URL-адреса буде: `/files//home/johndoe/myfile.txt`, з подвійною косою рискою (`//`) між `files` та `home`.

## Підсумок

З **FastAPI**, використовуючи короткі, інтуїтивно зрозумілі та стандартні оголошення типу Python, ви отримуєте:

> * Підтримка редактора: перевірка помилок, автозаповнення, тощо
> * "<abbr title="converting the string that comes from an HTTP request into Python data">Парсинг</abbr>" даних
> * Валідація даних
> * Анотація API та автоматична документація

Reviewer: ```suggestion
* Підтримку редактора: перевірка помилок, автозаповнення, тощо
* "<abbr title="converting the string that comes from an HTTP request into Python data">Парсинг</abbr>" даних
* Валідацію даних
* Анотації API та автоматична документацію
```
